src/live_restore/live_restore_fs.c

```diff
@@ -151,12 +151,16 @@ __live_restore_fs_create_stop_file(
     lr_fs = (WTI_LIVE_RESTORE_FS *)fs;
     path = path_marker = NULL;
 
-    /*
-     * FIXME-WT-14040: This is a big old race where we set this flag outside a lock and therefore
-     * have no idea if we will actually create a stop file after we clear the stop files.
-     */
-    if (lr_fs->finished)
+    bool reentrant = __wt_spin_owned(session, &lr_fs->state_lock);
+    if (!reentrant)
+        __wt_spin_lock(session, &lr_fs->state_lock);
+
+    WTI_LIVE_RESTORE_STATE state = __wti_live_restore_get_state(session, lr_fs);
+    if (WTI_LIVE_RESTORE_MIGRATION_COMPLETE(state)) {
+        if (!reentrant)
```

Comment:
There is a __wt_spin_unlock_if_owned function. Though that is vaguely different to what you are doing here. How does this path become re-entrant?

Reply:
> How does this path become re-entrant?

When we call `set_state` (taking the lock) we need to write the turtle file, and the `.turtle.set` rename creates needs a stop file.
I don't think `__wt_spin_unlock_if_owned` will help us here. Really we want the opposite of `unlock_if_taken_in_this_function`

Reply:
There are similar uses in `block/block_write.c:__block_write_off` where we have an incoming boolean of `caller_locked` and a local boolean of `local_locked`. That seems to be the same pattern you want here too.

Reply:
Could we move this to a follow on ticket? It's worth pursuing but I'd like to do so in a separate, smaller PR

Reply:
sgtm

Reply:
WT-14176 has been created to track this